### PR TITLE
fix(swebench-verified): eval pipeline fixes (meta-agent iter 1-28)

### PR DESCRIPTION
**Pull request by recursix**
Depends-on: cube-standard/fix/local-backend-pull-retry

## Summary

SWEBench eval pipeline fixes found by the meta-agent debug loop (iters 1–28), plus a new unified recipe replacing the three old `hello_swebench_*.py` scripts.

The `Depends-on` above carries one fix from cube-standard: Podman sets `DOCKER_HOST=http+unix://…` but the Docker Python SDK only accepts `unix://`. The normalization now lives in `LocalDockerServiceHandle` where the Docker client is created, invisible to recipes and users.

### `cubes/swebench-verified-cube/`
**`task.py`**
- `_apply_patch()` — remove hardcoded `cd /testbed` (broke after `relocate_if_readonly`); add `--forward` to prevent reverting already-applied patches
- `_run_tests()` — call `container.exec()` directly (exit code as int, no sentinel trick); drop `tail -n 200` sentinel
- `_build_test_cmd()` — remove `--no-header` (breaks pytest < 6.0); add `PYTHONIOENCODING=utf-8` for Django
- `evaluate()` — store full test output (no `[:2000]` slice)
- Remove `hint` field from `SWEBenchVerifiedTaskConfig` — hints go through `GennyConfig.task_hints`
- Remove manual `close()`; move `import re` to module level

**`benchmark.py`**
- Remove `_prefetch_images`, `_pull_with_retry`, `_log_docker_auth_status` — rate-limit retry moved to cube-standard
- `_setup()` simplified to 3 lines

**`tool.py`**
- `read_file`: coerce `line_start`/`line_end` to `int` — LLMs pass strings, causing `TypeError` at turn 17

**`debug.py`**
- Remove `oracle_mode` param — hardcode `True`; LLM can't find `/tmp/gold_patch.diff` anyway

### `cubes/swebench-live-cube/` + `cubes/terminalbench-cube/`
- Same `debug.py` cleanup: remove `oracle_mode` param

### `src/cube_harness/`
- `exp_runner.py`: fix `episode.config.task_id` → `episode.config.task_config.task_id` in 3 places (regression from 64f0b81)
- `agents/genny.py`: re-wrap all-tool obs groups as user messages — SWEBench obs are tool-only; stripping them left the agent blind (iter 29 fix)

### `recipes/swe_agent_recipe.py` (new — replaces `hello_swebench_*.py`)
- Unified recipe for swebench-verified, swebench-live, and terminalbench
- Lazy cube imports; argparse CLI (`--debug`, `--hints`, `--tasks`, `--subset`, `--n-parallel`, `--retry`)
- Uses `load_hints("swebench-verified")` from `meta_agent/hints/`

### `meta_agent/hints/swebench-verified.json`
- 21 task hints used via `GennyConfig.task_hints`

### `pyproject.toml`
- Add `docker>=7.1.0` explicit dependency

## Test plan

- [x] `psf__requests-1142` → reward=1.0 (gpt-5.4-mini)
- [x] `pallets__flask-5014` → reward=1.0
- [x] `make lint` passes
- [ ] Full swebench-verified run

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py [hunk 1]
```diff
@@ -118,30 +124,66 @@ def _apply_patch(self, patch: str) -> str:
...
+        test_directives: list[str],
+        timeout: int = 1800,
+        strict: bool = True,
+    ) -> tuple[bool, str]:
+        """Run test directives; return (all_passed, last-200-lines-of-output).
+
+        Output is trimmed to the last 200 lines because some repos (Django) print
+        tens of thousands of lines of DB-setup preamble before test results appear.
+
+        strict=False is used for pass_to_pass checks and relaxes two edge cases
```
Comment (recursix) [inline cubes/swebench-verified-cube/src/swebench_verified_cube/task.py:161]: can you discuss more strict=False? are we "modifying" swe bench here?

**Comment by recursix** [comment]
Re the `strict=False` comment on task.py:160 — short answer: no, we're not modifying SWE-bench's scoring, only patching around two data/container quality issues in the pass_to_pass suite that would penalise the agent unfairly.

**What strict=False does** (applies only to *pass_to_pass*, never fail_to_pass):

1. **Exit code 4 — no tests collected**: Some SWE-bench pass_to_pass entries contain truncated parametrised IDs (e.g. `test_stem[png-w/` with a missing `]`). pytest cannot collect them — they're malformed. Treating that as a failure penalises the agent for a dataset quality issue.

2. **Non-zero exit, zero failures**: Affects old sympy containers on Python 3.9 where `collections.MutableMapping` deprecation causes pytest to exit non-zero even though every test passed. The agent didn't cause this.

**fail_to_pass stays strict=True** — that's the resolution signal and is never relaxed.

**Vs official SWE-bench**: The princeton-nlp evaluator sidesteps both cases because it uses harness-specific test runners and doesn't encounter the truncated-ID problem. We're calling pytest more directly, which exposes these edge cases.

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py [hunk 2]
```diff
@@ -166,12 +206,15 @@ def _build_test_cmd(repo: str, test_directives: list[str]) -> str:
...
-            return f"./tests/runtests.py --verbosity 2 {tests}"
+            # PYTHONIOENCODING=utf-8: Django's test runner emits Unicode characters
+            # (e.g. U+2026 ellipsis) that fail when the container locale is ASCII-only.
+            return f"PYTHONIOENCODING=utf-8 ./tests/runtests.py --verbosity 2 {tests}"
         if "sympy" in repo:
             tests = " ".join(shlex.quote(t) for t in test_directives)
             return f"bin/test -C --verbose {tests}"
         tests = " ".join(shlex.quote(t) for t in test_directives)
-        return f"python -m pytest --no-header -rN -p no:cacheprovider {tests}"
+        # --no-header requires pytest>=6.0; many SWE-bench containers ship older versions.
```
Comment (recursix) [inline cubes/swebench-verified-cube/src/swebench_verified_cube/task.py:217]: any idea why --no-heder was there in the first place

Comment (recursix) [reply cubes/swebench-verified-cube/src/swebench_verified_cube/task.py:217]: --no-header suppresses the pytest version/platform header (e.g. `platform linux -- Python 3.x.x, pytest-x.x.x`). It was added to reduce noise in the stored output. It was removed here because some older pytest versions in SWE-bench containers don't recognise it and exit with an error before running any tests at all — making the whole evaluation fail silently. Dropping it is the safe cross-version choice.

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py [hunk 3]
```diff
@@ -183,20 +226,27 @@ class SWEBenchVerifiedTaskConfig(TaskConfig):
...
+                "SWEBenchVerifiedTaskConfig.make() requires runtime_context['infra'] "
+                "(preferred) or a legacy container_backend."
+            )
 
         # Import here to avoid circular import (benchmark imports task)
         from swebench_verified_cube.benchmark import SWEBenchVerifiedBenchmark
 
         metadata = SWEBenchVerifiedBenchmark.task_metadata[self.task_id]
         exec_info = SWEBenchVerifiedBenchmark.load_task_execution_info(self.task_id)
+        if self.hint:
```
Comment (recursix) [inline cubes/swebench-verified-cube/src/swebench_verified_cube/task.py]: we need to remove hints here, they need to go through genny config

Comment (recursix) [reply cubes/swebench-verified-cube/src/swebench_verified_cube/task.py]: Done — removed `hint: str | None = None` from `SWEBenchVerifiedTaskConfig` and the `if self.hint:` injection in `make()`. Hints now flow exclusively through `GennyConfig.task_hints`.

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py [hunk 4]
```diff
@@ -189,8 +232,12 @@ def make(
         runtime_context: RuntimeContext | None = None,
         container_backend: ContainerBackend | None = None,
     ) -> SWEBenchVerifiedTask:
-        if container_backend is None:
-            raise ValueError("SWEBenchVerifiedTaskConfig.make() requires a container_backend")
+        has_infra = runtime_context is not None and "infra" in runtime_context
```
Comment (recursix) [inline cubes/swebench-verified-cube/src/swebench_verified_cube/task.py]: container backend is legacy and no need to be backward compatible.

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py [hunk 5]
```diff
@@ -49,6 +49,15 @@ class SWEBenchVerifiedTask(Task):
     validate_per_step: bool = False
     accept_agent_stop: bool = True
 
+    def _build_tool(self) -> None:
```
Comment (recursix) [inline cubes/swebench-verified-cube/src/swebench_verified_cube/task.py:52]: can you put a docstring 1-liner

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py (hunk 4 above)
Comment (recursix) [reply cubes/swebench-verified-cube/src/swebench_verified_cube/task.py]: Done — simplified to require `runtime_context['infra']` exclusively. Removed the `has_infra` / `container_backend` fallback and the unused `ContainerBackend` import.

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py (hunk 5 above)
Comment (recursix) [reply cubes/swebench-verified-cube/src/swebench_verified_cube/task.py:52]: Done.

File: src/cube_harness/exp_runner.py [hunk 6]
```diff
@@ -73,7 +73,7 @@ def run_episode(episode: Episode) -> Trajectory:
     exp.benchmark.setup()
     try:
         episodes = exp.get_episodes_to_run()
-        ref_to_id = {run_episode.remote(episode): episode.config.task_id for episode in episodes}
+        ref_to_id = {run_episode.remote(episode): episode.config.task_config.task_id for episode in episodes}
```
Comment (recursix) [inline src/cube_harness/exp_runner.py]: can we make sure this change is compatible with up coming PR or recently merged PR.

Comment (recursix) [reply src/cube_harness/exp_runner.py]: This exact fix was independently merged to main as PR #316 ("fix(exp_runner): correct task_id access path on EpisodeConfig"). PR #329's version is consistent with it — no conflict. PR #326 (benchmark-config split) rebuilds exp_runner.py on top of a base that already has `task_config.task_id`, so it's clean there too.

File: cubes/swebench-verified-cube/src/swebench_verified_cube/task.py (hunk 1 above)
Comment (recursix) [reply cubes/swebench-verified-cube/src/swebench_verified_cube/task.py:161]: No SWE-bench semantics change — `strict=False` only relaxes two edge cases that are not the agent's fault, both only on `pass_to_pass` (regressions):

1. **exit 4** (no tests collected): SWE-bench stores some parametrised test IDs in truncated form (e.g. `test_stem[png-w/` missing the closing `]`) that pytest can't parse. The benchmark data is malformed; penalising the agent for it is wrong.
2. **non-zero exit, zero failures**: old sympy containers on Python 3.9 raise `collections.MutableMapping` deprecation errors at import time, inflating the exit code even when every test passed. We guard this with a "some tests ran AND no tests failed" check.

`fail_to_pass` (the primary resolution criterion) always uses `strict=True`, so a task is never counted as resolved unless the target tests actually pass.
